gateway-server/src/main/java/org/apache/knox/gateway/websockets/ProxyWebSocketAdapter.java

```diff
@@ -198,12 +232,34 @@ public void onError(Throwable cause) {
 
       @Override
       public void onMessageText(String message, Object session) {
-        final RemoteEndpoint remote = getRemote();
-
         LOG.logMessage("[From Backend <---]" + message);
+        remoteLock.lock();
+        final RemoteEndpoint remote = getRemote();
+        if (remote == null) {
+          LOG.logMessage("Remote endpoint is null");
+          if (messageBuffer == null) {
+             messageBuffer = new ArrayList<String>();
+          }
+          if (messageBuffer.size() >= MAX_BUFFER_MESSAGE_COUNT) {
+            throw new RuntimeIOException("Remote is null and message buffer is full. Cannot buffer anymore ");
+          }
+          LOG.logMessage("Buffering message: " + message);
+          messageBuffer.add(message);
+          remoteLock.unlock();
```

Comment:
Please surround this logic with a try-finally and move the unlock invocation in the finally block.

Reply:
Done. Thanks